### PR TITLE
Add DataLocation model and update storage data relation to many-to-many

### PR DESCRIPTION
**Pull request by hadalin**

### CODE REVIEW

File: resolwe/flow/models/data.py [hunk 1]
```diff
@@ -430,14 +441,28 @@ def save(self, render_name=False, *args, **kwargs):  # pylint: disable=keyword-a
...
+            self._save(*args, **kwargs)
 
             # We can only save dependencies after the data object has been saved. This
             # is why a transaction block is needed and the save method must be called first.
             if create:
                 self.save_dependencies(self.input, self.process.input_schema)  # pylint: disable=no-member
                 self.create_entity()
 
+    @transaction.atomic
+    def _save(self, *args, **kwargs):
```
Comment (dblenkus) [inline resolwe/flow/models/data.py]: Why have you added a separate method for this? There is already a ton of stuff in existing `save`.

File: resolwe/flow/models/data.py [hunk 2]
```diff
@@ -486,3 +511,13 @@ class DataDependency(models.Model):
...
+
+
+class DataLocation(models.Model):
+    """Location data of the data object."""
+
+    #: data object
+    data = models.OneToOneField(Data, on_delete=models.CASCADE, primary_key=True, related_name='location')
+
+    #: location identifier bound to data object's id
+    location_id = models.CharField(max_length=255)
```
Comment (dblenkus) [inline resolwe/flow/models/data.py]: You can set a lower length, e.g. 30.

Comment (dblenkus) [reply resolwe/flow/models/data.py]: And we want to index this as it will be used in purge.

File: resolwe/flow/tests/test_models.py [hunk 3]
```diff
@@ -663,23 +663,6 @@ def test_save_storage_file(self):
         self.assertEqual(data.output['json_field'], storage.pk)
         self.assertEqual(storage.json, {'foo': 'bar'})
 
-    def test_delete_data(self):
```
[the remark this answers is not in the document]
Comment (hadalin) [reply resolwe/flow/tests/test_models.py]: No, `Storage`s are deleted in purge, similarly to unreferenced data object's directory. Many data objects can point to the same storage.

File: resolwe/flow/models/utils.py [hunk 4]
```diff
@@ -435,32 +435,33 @@ def add_dir_size(obj):
...
-                    add_file_size(obj)
-                    data_size += obj.get('total_size', 0)
-            elif field_schema['type'].startswith('basic:dir:'):
-                add_dir_size(value)
-                data_size += value.get('total_size', 0)
-            elif field_schema['type'].startswith('list:basic:dir:'):
-                for obj in value:
-                    add_dir_size(obj)
-                    data_size += obj.get('total_size', 0)
+    if data.pk:
```
[the remark this answers is not in the document]
Comment (hadalin) [reply resolwe/flow/models/utils.py]: It's a bugfix and an optimization. Data should be hydrated only if the model instance exists.

File: resolwe/flow/utils/purge.py [hunk 5]
```diff
@@ -136,8 +136,8 @@ def data_purge(data_ids=None, delete=False, verbosity=0):
                 except ValueError:
                     continue
 
-                # Check if a data object with the given identifier exists.
-                if not Data.objects.filter(pk=data_id).exists():
+                # Check if any data object references given identifier.
+                if not Data.objects.filter_by_location(str(data_id)):
```
[the remark this answers is not in the document]
Comment (hadalin) [reply resolwe/flow/utils/purge.py]: True, but that would require quite a significant refactoring. For now it's at least easier to find where we deal with location, by using location related methods. I'll add a `# TODO` comment.

File: resolwe/flow/models/data.py (hunk 1 above)
Comment (hadalin) [reply resolwe/flow/models/data.py]: This `_save` is the absolute minimum _save_. If you, for example, want to implement duplicate function, one would call `_save` inside it, because you don't need validation and whatnot, but you still need to create `DataLocation` instance.

File: resolwe/flow/models/data.py [hunk 6]
```diff
@@ -430,14 +441,28 @@ def save(self, render_name=False, *args, **kwargs):  # pylint: disable=keyword-a
...
-            super().save(*args, **kwargs)
+            self._save(*args, **kwargs)
 
             # We can only save dependencies after the data object has been saved. This
             # is why a transaction block is needed and the save method must be called first.
             if create:
                 self.save_dependencies(self.input, self.process.input_schema)  # pylint: disable=no-member
                 self.create_entity()
 
+    @transaction.atomic
```
[the remark this answers is not in the document]
Comment (hadalin) [reply resolwe/flow/models/data.py]: For now yes, but in the future we could call `_save` somewhere where the code is not inside atomic block. The developer might forget to wrap it.

**Comment by hadalin** [comment]
Any recommended way of accessing `location_id` from within the [executor](https://github.com/genialis/resolwe/blob/9408726bc5e609d23e99fd4afe860cacc79432db/resolwe/flow/executors/docker/run.py#L150)? Only data id is passed via the [run](https://github.com/genialis/resolwe/blob/9408726bc5e609d23e99fd4afe860cacc79432db/resolwe/flow/executors/__main__.py#L51) function. Is it ok to import `Data` model inside the executor (getting some `Apps aren't loaded yet` kind of errors when running tests).

File: resolwe/flow/models/data.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (dblenkus) [reply resolwe/flow/models/data.py]: I would agree with this for a public method, but a private one can/should be a lot more specific and lightweight. You have to take into account that a transaction inside inside a transaction creates another savepoint and adds a bit of overhead.

File: resolwe/flow/models/data.py (hunk 1 above)
Comment (dblenkus) [reply resolwe/flow/models/data.py]: This function is too specific to be used in duplication function as it sets data location to data it, which is only true when new data object is created.

File: resolwe/flow/models/utils.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (dblenkus) [reply resolwe/flow/models/utils.py]: So, it has to be in a separate commit.

File: resolwe/flow/utils/purge.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (dblenkus) [reply resolwe/flow/utils/purge.py]: This is an absolutely crucial functionality, as it can cause a massive data loss if not done right.

File: resolwe/flow/tests/test_models.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (dblenkus) [reply resolwe/flow/tests/test_models.py]: As discussed, this should be handled in a signal. If you call purge with data id after data object is deleted, the current code doesn't leave any left-overs. So we have to keep this functionality.

**Comment by hadalin** [comment]
Accessing `location_id` https://github.com/genialis/resolwe/pull/571#issuecomment-448625068 is resolved [here](https://github.com/genialis/resolwe/pull/571/commits/cf9ebf440c1de9aa5c46a6810e025fad09d786c5).

**Comment by dblenkus** [comment]
> Any recommended way of accessing location_id from within the executor?
I think it should already be present in settings passed to executor, so you can just use it from there.

File: resolwe/flow/tests/test_models.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (hadalin) [reply resolwe/flow/tests/test_models.py]: >If you call purge with data id after data object is deleted, the current code doesn't leave any left-overs.

Not entirely true. Data objects directories are [deleted only when purge is called without data ids](https://github.com/genialis/resolwe/blob/master/resolwe/flow/utils/purge.py#L126-L141).

My reasoning was to do the same with storages: leave it until purge without ids is called.

File: resolwe/flow/utils/purge.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (hadalin) [reply resolwe/flow/utils/purge.py]: No shit :)

File: resolwe/flow/tests/test_models.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (dblenkus) [reply resolwe/flow/tests/test_models.py]: > Data objects directories are deleted only when purge is called without data ids.

Not entirely true ;). You just have to add `data_purge(data_id)` in API after deleting data object and it works as described above.

File: resolwe/flow/utils/purge.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (hadalin) [reply resolwe/flow/utils/purge.py]: More tests it is.

File: resolwe/flow/models/data.py (hunk 1 above)
Comment (dblenkus) [reply resolwe/flow/models/data.py]: I agree that this split is useful, I just wanted to say that it doesn't support all use-cases, because it sets `location_id=str(self.pk)`.
And i am in favour of a more descriptive name instead of `_save` - maybe `perform_save`, to follow DRF's practice.

File: resolwe/flow/models/data.py [hunk 7]
```diff
@@ -486,3 +508,13 @@ class DataDependency(models.Model):
...
+
+
+class DataLocation(models.Model):
+    """Location data of the data object."""
+
+    #: data object
+    data = models.OneToOneField(Data, on_delete=models.CASCADE, primary_key=True, related_name='location')
+
+    #: location identifier bound to data object's id
+    location_id = models.CharField(max_length=30, db_index=True)
```
Comment (dblenkus) [inline resolwe/flow/models/data.py]: `location_id` -> `location`

File: docs/contributing.rst [hunk 8]
```diff
@@ -98,7 +98,7 @@ To run the tests with Tox_, use::
 
 .. _Tox: http://tox.testrun.org/
 
-Buildling documentation
+Building documentation
```
Comment (dblenkus) [inline docs/contributing.rst]: This should go into a separate commit - together with all other typos.

File: docs/CHANGELOG.rst [hunk 9]
```diff
@@ -42,6 +50,7 @@ Changed
 - Increase Docker's memory limit for 100MB to make sure processes are not
   killed when using all available memory and tune Docker memory limits to
   avoid OOM.
+  killed when using all available memory
```
Comment (dblenkus) [inline docs/CHANGELOG.rst]: This was accidentally added.

File: resolwe/flow/utils/purge.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (dblenkus) [reply resolwe/flow/utils/purge.py]: My first comment is related to the fact that variable is named `data_id` - which is not correct. If you look few lines higher, it is folder's name, so you should rename it to `location`. It assumes that location is an integer, but this is something we can live with for now.

File: resolwe/flow/models/data.py (hunk 1 above)
Comment (hadalin) [reply resolwe/flow/models/data.py]: >And i am in favour of a more descriptive name instead of _save - maybe perform_save, to follow DRF's practice.

That makes it a "public" function. Does it need to be decorated with `@transaction.atomic`?

File: resolwe/flow/models/data.py (hunk 7 above)
Comment (hadalin) [reply resolwe/flow/models/data.py]: Thinking about this. Even `location` doesn't sound right. Location can mean anything really. It could mean absolute "physical" path, "https" path. What about `directory_name`?

Comment (hadalin) [reply resolwe/flow/models/data.py]: That way [this](https://github.com/genialis/resolwe/pull/571/files#diff-1adac73ebf52d60a1bde30c2fceb4610R136) can be more verbose.

Comment (dblenkus) [reply resolwe/flow/models/data.py]: This is exactly the point here, at some point we will add S3-like storage or something similar and this model has also to cover that case. It can be `location`, `path` or something similar. `location_id` is a bad choice because it is too similar to `location.id`.

File: resolwe/flow/models/data.py (hunk 1 above)
Comment (dblenkus) [reply resolwe/flow/models/data.py]: Good question :).

If a Data object is created over the API, there are at least three transactions  ([here](https://github.com/genialis/resolwe/blob/master/resolwe/flow/views/data.py#L174), [here](https://github.com/genialis/resolwe/blob/master/resolwe/flow/models/data.py#L432) and this one). And there are still lines like [this one](https://github.com/genialis/resolwe/blob/master/resolwe/flow/models/data.py#L398) that should be in s transaction, but are not always (e.g. when executor updates the output).

A good solution nay be to name the method `_perform_create` and remove the transaction.

**Comment by hadalin** [comment]
Ready for another review round.

File: resolwe/flow/models/data.py [hunk 10]
```diff
@@ -486,3 +487,28 @@ class DataDependency(models.Model):
     parent = models.ForeignKey(Data, null=True, on_delete=models.SET_NULL, related_name='children_dependency')
     #: kind of dependency
     kind = models.CharField(max_length=16, choices=KIND_CHOICES)
+
+
+class DataLocation(models.Model):
+    """Location data of the data object."""
+
+    #: location (bound to data object's id)
+    location = models.CharField(max_length=30, db_index=True)
```
Comment (hadalin) [inline resolwe/flow/models/data.py]: Should `location` be unique?

File: resolwe/flow/models/utils.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (hadalin) [reply resolwe/flow/models/utils.py]: I don't have any numbers, but one reason for this change was because tests failed due to location not being created yet.

File: resolwe/flow/tests/test_models.py [hunk 11]
```diff
@@ -155,11 +154,15 @@ def test_hydrate_file_size(self):
             'output_file': {'file': 'output.txt'}
         }
 
+        data_location = DataLocation.objects.create(location=str(data.pk))
```
[the remark this answers is not in the document]
Comment (hadalin) [reply resolwe/flow/tests/test_models.py]: The `location` does not play a significant part here as long as it's unique. It's less code that way which means less complexity.

File: resolwe/flow/tests/test_purge.py [hunk 12]
```diff
@@ -352,78 +360,117 @@ def tearDown(self):
...
             os_mock.remove = MagicMock()
-            purge.data_purge()
+            purge.purge_all()
             os_mock.remove.assert_not_called()
 
         # Check that only the 'removeme' file from the completed Data objects is removed
         # and files from the second (not completed) Data objects are unchanged.
         with patch('resolwe.flow.utils.purge.os', wraps=os) as os_mock:
-            os_mock.path.isfile = MagicMock(return_value=True)
+            # os_mock.path.isfile = MagicMock(return_value=True)
```
[the remark this answers is not in the document]
Comment (hadalin) [reply resolwe/flow/tests/test_purge.py]: This returns `True` no matter what. It's better for `isfile` to return the real value. Forgot to remove the commented out code.

[the remark this answers is not in the document]
Comment (dblenkus) [reply resolwe/flow/tests/test_purge.py]: I agree it is better to do it without mock if it works. But then change it all over the test.

File: resolwe/flow/tests/test_models.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (dblenkus) [reply resolwe/flow/tests/test_models.py]: But it can imply a wrong thing.

File: resolwe/flow/models/utils.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (dblenkus) [reply resolwe/flow/models/utils.py]: It looks like tests are doing it in a wrong sequence. `output` should not be populated if `DataLocation` doesn't exist. If `output` is empty the for loop should never execute as `iterate_fields` returns an empty list.

File: resolwe/flow/models/data.py [hunk 13]
```diff
@@ -486,3 +487,28 @@ class DataDependency(models.Model):
...
+        if filename:
+            path = os.path.join(path, filename)
+
+        return path
+
+    def get_runtime_path(self, filename=None):
+        """Compose data runtime location path."""
+        return self.get_path(prefix=settings.FLOW_EXECUTOR['RUNTIME_DIR'], filename=filename)
+
+    def get_upload_path(self, filename=None):
```
[the remark this answers is not in the document]
Comment (hadalin) [reply resolwe/flow/models/data.py]: Isn't it bound to location, here `path = os.path.join(prefix, self.location)`?

[the remark this answers is not in the document]
Comment (dblenkus) [reply resolwe/flow/models/data.py]: No, location is not a part of upload path. It is assembled just from `FLOW_EXECUTOR['UPLOAD_DIR']` and `filename`.

File: resolwe/flow/migrations/0026_add_data_location.py [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+        with transaction.atomic():
+            # Manually set DataLocation id to preserve
+            # data directory.
+            #
+            # Before migration, FLOW_EXECUTOR['DATA_DIR'] contained directories
+            # named by data object IDs.
+            # After migration, FLOW_EXECUTOR['DATA_DIR'] contains directories
+            # named by DataLocation IDs.
+            data_location = DataLocation.objects.create(location='')
+            data_location.location = str(data_location.id)
```
[the remark this answers is not in the document]
Comment (hadalin) [reply resolwe/flow/migrations/0026_add_data_location.py]: Lol, I had this, but overwritten myself. The comment stayed `# Manually set DataLocation id to preserve data directory.`

File: resolwe/flow/models/data.py [hunk 15]
```diff
@@ -486,3 +487,28 @@ class DataDependency(models.Model):
...
+        prefix = prefix or settings.FLOW_EXECUTOR['DATA_DIR']
+
+        path = os.path.join(prefix, self.location)
+        if filename:
+            path = os.path.join(path, filename)
+
+        return path
+
+    def get_runtime_path(self, filename=None):
+        """Compose data runtime location path."""
```
[the remark this answers is not in the document]
Comment (hadalin) [reply resolwe/flow/models/data.py]: It's supported in `get_path`. This method just passes the argument.

File: resolwe/flow/utils/purge.py [hunk 16]
```diff
@@ -97,57 +98,52 @@ def subfiles(root):
...
-                    continue
-
-                # Check if a data object with the given identifier exists.
-                if not Data.objects.filter(pk=data_id).exists():
-                    unreferenced_files.add(directory_path)
+    unreferenced_files = set()
+    is_location_referenced_by_data = location.data.exists()
+    if is_location_referenced_by_data:
+        # Perform cleanup.
+        only_ok_or_error = location.data.exclude(Q(status=Data.STATUS_DONE) | Q(status=Data.STATUS_ERROR)).count() == 0
```
[the remark this answers is not in the document]
Comment (hadalin) [reply resolwe/flow/utils/purge.py]: This assumes there's going to be periods of no data processing. What if the server is "very active" and incessantly processing? Is `purge` called only on downtime?

File: resolwe/flow/utils/purge.py [hunk 17]
```diff
@@ -97,57 +98,52 @@ def subfiles(root):
...
-                try:
-                    data_id = int(directory)
-                except ValueError:
-                    continue
-
-                # Check if a data object with the given identifier exists.
-                if not Data.objects.filter(pk=data_id).exists():
-                    unreferenced_files.add(directory_path)
+    unreferenced_files = set()
+    is_location_referenced_by_data = location.data.exists()
```
[the remark this answers is not in the document]
Comment (hadalin) [reply resolwe/flow/utils/purge.py]: The `is_location_referenced_by_data` is used in two places and longer names adds to readability. No?

[the remark this answers is not in the document]
Comment (dblenkus) [reply resolwe/flow/utils/purge.py]: Longer, not too long :). But each variable also adds some overhead as you have to figure out where it comes from. In this case i would argue to use `location.data.exists()` in both cases, especially as it is shorter as `is_location_referenced_by_data`.

[the remark this answers is not in the document]
Comment (hadalin) [reply resolwe/flow/utils/purge.py]: Using a variable with a meaningful name saves you from writing comments. You also pay extra attention when changing functionality because it forces you look where that variable is used.

File: resolwe/flow/tests/test_purge.py [hunk 18]
```diff
@@ -474,3 +474,47 @@ def test_remove_same_location(self):
...
+        data_2.output = {'sample': {'file': 'test-file'}}
+        self.create_test_file(data_2.location, 'test-file')
+        data_2.storages.add(storage_1)
+        data_2.save()
+
+        self.assertEqual(Data.objects.count(), 2)
+        self.assertEqual(Storage.objects.count(), 2)
+
+        # Delete first object.
+        data_1.delete()
```
[the remark this answers is not in the document]
Comment (dblenkus) [reply resolwe/flow/tests/test_purge.py]: I was thinking about something like:

```
def delete(...):
    storages = list(self.storage.all())
    super().delete(...)
    for storage in storages:
        <delete if orphaned>
```

File: resolwe/flow/utils/purge.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (dblenkus) [reply resolwe/flow/utils/purge.py]: Sorry, i meant `if location.data.exclude...: return`

File: resolwe/flow/tests/test_purge.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (hadalin) [reply resolwe/flow/tests/test_purge.py]: Doesn't this have a significant performance impact? Could this produce a timeout?

File: resolwe/flow/models/utils.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (hadalin) [reply resolwe/flow/models/utils.py]: This method calculates file and directory sizes. If data objects is not yet created, it also means is does not have a location so there's not point in continuing. I changed the if statement to `if data.location:` which makes the if statement self-explanatory.

**Comment by hadalin** [comment]
Addressed your comments. [This](https://github.com/genialis/resolwe/pull/571#discussion_r241381269), [this](https://github.com/genialis/resolwe/pull/571#discussion_r249302083), and [this](https://github.com/genialis/resolwe/pull/571#discussion_r249440801) comments are unresolved.

File: resolwe/flow/managers/dispatcher.py [hunk 19]
```diff
@@ -404,19 +404,25 @@ def _get_per_data_dir(self, dir_base, data_id):
...
+        logger.debug(__("Preparing data directory for Data with id {}.", data))
+
+        with transaction.atomic():
+            data_location = DataLocation.objects.create(location='')
+            data_location.location = str(data_location.id)
+            data_location.save()
+            data_location.data.add(data)
 
-        output_path = self._get_per_data_dir('DATA_DIR', data_id)
+        output_path = self._get_per_data_dir('DATA_DIR', data_location.location)
```
[the remark this answers is not in the document]
Comment (dblenkus) [reply resolwe/flow/managers/dispatcher.py]: You are right. Leave this as is. And also notice that this comment implies that you shouldn't use `DataLocation.get_*_path` in dispatcher.

File: resolwe/flow/utils/purge.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (dblenkus) [reply resolwe/flow/utils/purge.py]: For me `location.data.exists()` is more meaningful than `is_location_referenced_by_data` if I only see a chunk of code.
I won't argue to remove it if you prefer it this way (it saves one db query, so it is better after all). But please find a shorter name :)

File: resolwe/flow/models/utils.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (dblenkus) [reply resolwe/flow/models/utils.py]: I agree that `data.location` is much better. But file and directory sizes are retrieved from `data.output`. So it would make even more sense to use that - but note that this condition is already (implicitly) included in the following `for`, if `data.output` is empty, the loop will never execute.

File: resolwe/flow/tests/test_purge.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (hadalin) [reply resolwe/flow/tests/test_purge.py]: Does it even make sense to have storage purge then?

[the remark this answers is not in the document]
Comment (hadalin) [reply resolwe/flow/tests/test_purge.py]: >Does it even make sense to have storage purge then?

Oh, it does because there's other ways for data object to be deleted (instead via `Data.delete()`).

**Comment by hadalin** [comment]
- https://github.com/genialis/resolwe/pull/571#discussion_r241381269 Removed "hydrate optimization" (`if data.pk`) and fixed the [tests](https://github.com/genialis/resolwe/pull/571/files#diff-33c302481d283000e7573635e29d1c86R541).
- https://github.com/genialis/resolwe/pull/571#discussion_r249440801 Fixed [storage](https://github.com/genialis/resolwe/pull/571/files#diff-fe592d277a39630035897e67d6967343R498) [deletion](https://github.com/genialis/resolwe/pull/571/files#diff-aa73ad8a382e284ba1af2170c5e136caR440).

File: resolwe/flow/managers/dispatcher.py [hunk 20]
```diff
@@ -404,19 +405,29 @@ def _get_per_data_dir(self, dir_base, data_id):
...
         """
-        logger.debug(__("Preparing data directory for Data with id {}.", data_id))
-
-        output_path = self._get_per_data_dir('DATA_DIR', data_id)
+        logger.debug(__("Preparing data directory for Data with id {}.", data.id))
+
+        with transaction.atomic():
+            # Create a temporary random location
+            # and then override it with data location id
+            # since object have to be created first.
```
Comment (mstajdohar) [inline resolwe/flow/managers/dispatcher.py]: have -> has

File: resolwe/flow/models/data.py [hunk 21]
```diff
@@ -486,3 +487,24 @@ class DataDependency(models.Model):
     parent = models.ForeignKey(Data, null=True, on_delete=models.SET_NULL, related_name='children_dependency')
     #: kind of dependency
     kind = models.CharField(max_length=16, choices=KIND_CHOICES)
+
+
+class DataLocation(models.Model):
+    """Location data of the data object."""
+
+    #: location (bound to data object's id)
+    location = models.CharField(max_length=30, unique=True)
```
Comment (mstajdohar) [inline resolwe/flow/models/data.py]: The notation `data.location.location` is a bit cumbersome. Could you name this differently? What exactly do you store in this field? Could we call it `slug` or `segment`?

File: resolwe/flow/utils/purge.py [hunk 22]
```diff
@@ -97,57 +96,52 @@ def subfiles(root):
...
-    """Print files not referenced from meta data.
-
-    If data_ids not given, run on all data objects.
-    If delete is True, delete unreferenced files.
+def location_purge(location_id, delete=False, verbosity=0):
+    """Print and conditionally delete files not referenced by meta data.
 
+    :param location_id: Id of the
+        :class:`~resolwe.flow.models.DataLocation` data objects
+        reference to.
```
Comment (mstajdohar) [inline resolwe/flow/utils/purge.py]: typo?

File: resolwe/flow/utils/purge.py [hunk 23]
```diff
@@ -97,57 +96,52 @@ def subfiles(root):
...
+                data.output,
+                data.process.output_schema,
+                data.descriptor,
+                getattr(data.descriptor_schema, 'schema', [])
+            ))
+
+        intersected_files = reduce(lambda acc, cur: acc.intersection(cur), purge_files_sets)
+        unreferenced_files.update(intersected_files)
+    else:
+        # Perform emptying.
```
Comment (mstajdohar) [inline resolwe/flow/utils/purge.py]: Consider changing the comment to Remove data directory

File: resolwe/flow/models/data.py (hunk 21 above)
Comment (dblenkus) [reply resolwe/flow/models/data.py]: I suggest `path`, or maybe `subpath`, as in future there will be multiple storage types (local storage, s3,...) and this will be a subpath of the storage path.

Comment (hadalin) [reply resolwe/flow/models/data.py]: `subpath` it is.

File: resolwe/flow/migrations/0026_add_data_location.py [hunk 24]
```diff
@@ -0,0 +1,42 @@
...
+
+def set_data_location(apps, schema_editor):
+    """Create DataLocation for each Data."""
+    Data = apps.get_model('flow', 'Data')
+    DataLocation = apps.get_model('flow', 'DataLocation')
+
+    for data in Data.objects.all():
+        with transaction.atomic():
+            # Manually set DataLocation id to preserve
+            # data directory.
```
Comment (dblenkus) [inline resolwe/flow/migrations/0026_add_data_location.py]: Put the comment in one line.

File: resolwe/flow/managers/dispatcher.py [hunk 25]
```diff
@@ -574,6 +585,7 @@ def override_settings(self, **kwargs):
         elif cmd == WorkerProtocol.FINISH:
             try:
                 data_id = message[WorkerProtocol.DATA_ID]
+                data_location = Data.objects.get(id=data_id).location
```
Comment (dblenkus) [inline resolwe/flow/managers/dispatcher.py]: Do the query with `DataLocation` instead of `Location` as it is more efficient (uses less joins).

File: resolwe/flow/executors/docker/prepare.py [hunk 26]
```diff
@@ -37,10 +37,7 @@ def resolve_data_path(self, data=None, filename=None):
         if data is None:
             return constants.DATA_ALL_VOLUME
 
-        if filename is None:
-            return os.path.join(constants.DATA_ALL_VOLUME, str(data.id))
-
-        return os.path.join(constants.DATA_ALL_VOLUME, str(data.id), filename)
+        return data.location.get_path(prefix=constants.DATA_ALL_VOLUME, filename=filename)
```
[the remark this answers is not in the document]
Comment (hadalin) [reply resolwe/flow/executors/docker/prepare.py]: Done, but it might be confusing because a few lines above you have 

```py
if not getattr(settings, 'FLOW_DOCKER_DONT_PULL', False):
```

where `settings` are `from django.conf import settings`. Might be worth adding a comment there too.

File: resolwe/flow/executors/prepare.py [hunk 27]
```diff
@@ -84,10 +85,7 @@ def resolve_data_path(self, data=None, filename=None):
         if data is None:
             return settings.FLOW_EXECUTOR['DATA_DIR']
 
-        if filename is None:
-            return os.path.join(settings.FLOW_EXECUTOR['DATA_DIR'], str(data.id))
-
-        return os.path.join(settings.FLOW_EXECUTOR['DATA_DIR'], str(data.id), filename)
+        return data.location.get_path(filename=filename)
```
[the remark this answers is not in the document]
Comment (hadalin) [reply resolwe/flow/executors/prepare.py]: Throughout this file Django's settings are used. It's not immediately obvious when one should and shouldn't use Django's settings directly.

[the remark this answers is not in the document]
Comment (hadalin) [reply resolwe/flow/executors/prepare.py]: I assume `from .global_settings import DJANGO_SETTINGS` must be used?

File: resolwe/flow/managers/dispatcher.py [hunk 28]
```diff
@@ -495,10 +506,10 @@ def _prepare_context(self, data_id, data_dir, runtime_dir, **kwargs):
             finally:
                 os.umask(old_umask)
 
-    def _prepare_executor(self, data_id, executor):
+    def _prepare_executor(self, data, executor):
         """Copy executor sources into the destination directory.
 
-        :param data_id: The :class:`~resolwe.flow.models.Data` object id
+        :param data: The :class:`~resolwe.flow.models.Data` object
```
[the remark this answers is not in the document]
Comment (hadalin) [reply resolwe/flow/managers/dispatcher.py]: Ok, but putting `being` into this line makes it 73 characters long.

File: resolwe/flow/managers/dispatcher.py [hunk 29]
```diff
@@ -28,7 +29,8 @@
 
 from resolwe.flow.engine import InvalidEngineError, load_engines
 from resolwe.flow.execution_engines import ExecutionError
-from resolwe.flow.models import Data, DataDependency, Process
+from resolwe.flow.models import Data, DataDependency, DataLocation
+from resolwe.flow.models.process import Process
```
[the remark this answers is not in the document]
Comment (hadalin) [reply resolwe/flow/managers/dispatcher.py]: Lapsus.

File: resolwe/flow/utils/purge.py [hunk 30]
```diff
@@ -164,6 +164,24 @@ def _location_purge_all(delete=False, verbosity=0):
         logger.info("No data locations")
 
 
+def storage_purge(delete=False, verbosity=0):
```
Comment (dblenkus) [inline resolwe/flow/utils/purge.py]: It is not intuitive that `location_purge` purges single location and `storage_purge` purges all storages.
I don't see a use-case where this function would be used separately. So you can just rename it to `_storage_purge_all` and only use it in `purge_all`.

File: resolwe/flow/utils/purge.py [hunk 31]
```diff
@@ -164,6 +164,24 @@ def _location_purge_all(delete=False, verbosity=0):
...
+    """Purge unreferenced storages."""
+    orphaned_storages = Storage.objects.filter(data=None)
+
+    if verbosity >= 1:
+        if orphaned_storages is not None:
+            if orphaned_storages:
```
Comment (dblenkus) [inline resolwe/flow/utils/purge.py]: Can you combine this `if` with the one above?

**Comment by dblenkus** [comment]
And note that purge has changed on the master branch, so pay attention when rebasing.

File: resolwe/flow/managers/dispatcher.py [hunk 32]
```diff
@@ -574,6 +584,7 @@ def override_settings(self, **kwargs):
         elif cmd == WorkerProtocol.FINISH:
             try:
                 data_id = message[WorkerProtocol.DATA_ID]
+                data_location = DataLocation.objects.get(data__id=data_id)
```
[the remark this answers is not in the document]
Comment (hadalin) [reply resolwe/flow/managers/dispatcher.py]: You mean `Data.objects.get(id=data_id).location.first()`? I think it was this way before, but we changed it.

File: resolwe/flow/tests/test_models.py [hunk 33]
```diff
@@ -531,45 +537,65 @@ def test_process_isactive(self):
...
+        data.output = {'test_file': {'file': 'test_file.tmp'}}
+        data.save()
+
+        return data
+
     def setUp(self):
         super().setUp()
 
-        self.process = Process(
+        self.process = Process.objects.create(
```
[the remark this answers is not in the document]
Comment (hadalin) [reply resolwe/flow/tests/test_models.py]: Because otherwise test throws `ValueError: save() prohibited to prevent data loss due to unsaved related object 'process'.`

File: resolwe/flow/utils/purge.py [hunk 34]
```diff
@@ -97,57 +96,52 @@ def subfiles(root):
...
-def data_purge(data_ids=None, delete=False, verbosity=0):
-    """Print files not referenced from meta data.
-
-    If data_ids not given, run on all data objects.
-    If delete is True, delete unreferenced files.
+def location_purge(location_id, delete=False, verbosity=0):
+    """Print and conditionally delete files not referenced by meta data.
 
+    :param location_id: Id of the
+        :class:`~resolwe.flow.models.DataLocation` model that
```
[the remark this answers is not in the document]
Comment (hadalin) [reply resolwe/flow/utils/purge.py]: 72 is mid-word. Did it as 67.

File: resolwe/flow/managers/dispatcher.py [hunk 35]
```diff
@@ -413,19 +413,29 @@ def _get_per_data_dir(self, dir_base, data_id):
...
         """
-        logger.debug(__("Preparing data directory for Data with id {}.", data_id))
-
-        output_path = self._get_per_data_dir('DATA_DIR', data_id)
+        logger.debug(__("Preparing data directory for Data with id {}.", data.id))
+
+        with transaction.atomic():
+            # Create a temporary random location and then override it with data
+            # location id since object has to be created first.
+            # TODO Find a better solution than creating a temporary location.
```
Comment (dblenkus) [inline resolwe/flow/managers/dispatcher.py]: `# TODO Find a better solution, e.g. defer the database constraint.`

File: resolwe/flow/managers/listener.py [hunk 36]
```diff
@@ -485,13 +485,13 @@ def handle_finish(self, obj):
...
                             'type': TYPE_PURGE_RUN,
-                            'data_id': data_id,
+                            'location_id': d.location.id,
                             'verbosity': self._verbosity,
                         }
                     )
                 except ChannelFull:
                     logger.warning(
-                        "Cannot triger purge because channel is full.",
+                        "Cannot trigger purge because channel is full.",
```
Comment (dblenkus) [inline resolwe/flow/managers/listener.py]: This should go into the previous commit.

File: resolwe/flow/tests/test_models.py [hunk 37]
```diff
@@ -645,6 +675,10 @@ def test_save_storage_file(self):
             contributor=self.contributor,
             process=self.proc,
         )
+        data_location = DataLocation.objects.create(subpath='')
```
Comment (dblenkus) [inline resolwe/flow/tests/test_models.py]: Is data location required for testing storage?

File: resolwe/flow/tests/test_models.py [hunk 38]
```diff
@@ -766,6 +800,10 @@ def test_hydrate_input_references(self):
             },
             size=0,
         )
+        data_location = DataLocation.objects.create(subpath='')
```
Comment (dblenkus) [inline resolwe/flow/tests/test_models.py]: You should use a non-empty path for testing that path is rendered correctly.

File: resolwe/flow/tests/test_models.py (hunk 37 above)
Comment (hadalin) [reply resolwe/flow/tests/test_models.py]: Yes, somewhere along the way `location.get_path` gets called.

File: resolwe/flow/tests/test_models.py (hunk 38 above)
Comment (hadalin) [reply resolwe/flow/tests/test_models.py]: The path gets set one line below.
